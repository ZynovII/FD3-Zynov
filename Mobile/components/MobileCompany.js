import React, { Fragment } from 'react';

import { EditTableEvents } from './events';
import CompanyTitle from './CompanyTitle';
import ClientsTable from './ClientsTable';
import ClientCardEdit from './ClientCardEdit';

import './MobileCompany.scss';

const fetch = typeof window !== 'undefined' ? window.fetch : require('node-fetch');

class MobileCompany extends React.PureComponent {

  state = {
    dataReady: true,
    clientsArr : this.props.clientsArr,
    filteredClientsArr: null,
    filterMode: 0, //0-standart view, 1-when it filtered
    companyes : ['Velcom', 'MTS'],
    selectedClient: null, //selected client id
    cardmode: 0, //0-no card, 1-card edit, 2-card add
    isValid: true,
  }

  componentDidMount = () => {
      EditTableEvents.addListener('EDeliteClient', this.deliteClient);
      EditTableEvents.addListener('EEditClient', this.editClient);
      EditTableEvents.addListener('ESetValid', this.setValid);
      EditTableEvents.addListener('ESaveChanges', this.saveClientChanges);
      EditTableEvents.addListener('ECancelChanges', this.cancelChanges);
      //this.loadData();
  };

  componentWillUnmount = () => {
      EditTableEvents.removeListener('EDeliteClient', this.deliteClient);
      EditTableEvents.removeListener('EEditClient', this.editClient);
      EditTableEvents.removeListener('ESetValid', this.setValid);
      EditTableEvents.removeListener('ESaveChanges', this.saveClientChanges);
      EditTableEvents.removeListener('ECancelChanges', this.cancelChanges);
  };

  loadData = () => {
    fetch( 'http://localhost:3000/' )
    .then( (response) => response.json() )
    .then( (result) => this.setState( {clientsArr: result, dataReady: true} ));
  }

  deliteClient = (id) => {
    let newClientsArr = this.state.clientsArr.filter( v => v.id != id);
    this.setState({clientsArr: newClientsArr, selectedClient: null, cardmode: 0});
  }

  editClient = (id) => {
    this.setState({cardmode: 1, selectedClient: id});
  }
  
  setValid = (bool) => this.setState( {isValid: bool} );

  addNewClient = () => {
    this.setState({cardmode: 2, selectedClient: null});
  }

  saveClientChanges = (client) => {
    let newClientArr = this.state.clientsArr.map(i => i.id == client.id ? client : i);
        if(client.id !== this.state.selectedClient) {
            newClientArr.push(client);
        }
        this.setState({clientsArr: newClientArr, cardmode: 0});
  }

  cancelChanges = () => this.setState({cardmode: 0, selectedClient: 0, isValid: true});

  showAllClients = () => this.setState( {filterMode: 0} );
  showActiveClients = () => {
    let newList = this.state.clientsArr.filter( v => v.status == 'active')
    this.setState( { filteredClientsArr: newList, filterMode: 1 } );
  };
  showBlockedClients = () => {
    let newList = this.state.clientsArr.filter( v => v.status == 'blocked')
    this.setState( { filteredClientsArr: newList, filterMode: 1 } );
  };

  render() {
    if(this.state.dataReady) {
      let newClientId = this.state.clientsArr[this.state.clientsArr.length-1].id + 1;
      var selectedClient = this.state.clientsArr.find( (v) => v.id === this.state.selectedClient ) ||
      {id: newClientId};
    }
    console.log('Компонент MobileCompany render');
    return (
      <div className='mobile-company'>
        <CompanyTitle companyes={this.state.companyes}/>
        <hr />
        <button dataTestId='filterAll' onClick={this.showAllClients}>All</button>
        <button dataTestId='filterActiv' onClick={this.showActiveClients}>Active</button>
        <button dataTestId='filterBlock' onClick={this.showBlockedClients}>Blocked</button>
        <hr />
        {
          this.state.cardmode !==0 && 
          <ClientCardEdit 
              key={selectedClient.id}
              client={selectedClient}
              cardmode={this.state.cardmode}
          />
        }
        {
          !this.state.dataReady 
          ? <div className='loading'>Data loading...</div>
          : [
              <ClientsTable key={'a'} users={ !this.state.filterMode
                ? this.state.clientsArr 
                : this.state.filteredClientsArr }
              />,
              <button dataTestId='add' key={'b'} onClick={this.addNewClient}>Add client</button>
            ]
        }
      </div>
    );
  }

}

export default MobileCompany;
