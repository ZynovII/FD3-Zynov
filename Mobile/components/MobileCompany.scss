@import '_main';

.mobile-company {
  font-size: $fontSize;
  .loading{
    font-size: 40px;
    font-weight: 600;
    animation: 4s linear 0s infinite alternate loading;
  }
  button {
    @extend %tplBtn;
  }
}

@keyframes loading {
  from {color: black;}
  to {color: seagreen};
}