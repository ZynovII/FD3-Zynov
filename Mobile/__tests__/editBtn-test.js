import React from 'react';
import renderer from 'react-test-renderer';

import MobileCompany from '../components/MobileCompany';

test('edit buttons work', () => {

  const component = renderer.create(
    <MobileCompany clientsArr={[ 
      {id:1, firstName:"Иван", lastName: "Иванов", patronymic: "Иванович", balance:200, status: 'active'}, 
      {id:2, firstName:"Сидор", lastName: "Сидоров", patronymic: "Сидорович", balance:250, status: 'active'}, 
      {id:3, firstName:"Петр", lastName: "Петров", patronymic: "Петрович", balance:180, status: 'active'},
      {id:4, firstName:"Григорий", lastName: "Григорьев", patronymic: "Григорьевич", balance:-220, status: 'blocked'},
    ]}/>
  );
  const editBtn = component.root.find( el => el.props.dataTestId == 'edit1');

  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  editBtn.props.onClick();
  
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
});