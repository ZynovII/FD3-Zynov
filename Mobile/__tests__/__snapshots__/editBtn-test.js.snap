// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit buttons work 1`] = `
<div
  className="mobile-company"
>
  <div>
    <button
      onClick={[Function]}
    >
      Velcom
    </button>
    <button
      onClick={[Function]}
    >
      MTS
    </button>
    <br />
    Company: 
    <span>
      Velcom
    </span>
  </div>
  <hr />
  <button
    dataTestId="filterAll"
    onClick={[Function]}
  >
    All
  </button>
  <button
    dataTestId="filterActiv"
    onClick={[Function]}
  >
    Active
  </button>
  <button
    dataTestId="filterBlock"
    onClick={[Function]}
  >
    Blocked
  </button>
  <hr />
  <table
    className="clients-table"
  >
    <tbody>
      <tr
        className="table__head"
      >
        <td>
          First Name
        </td>
        <td>
          Last Name
        </td>
        <td>
          Patronymic
        </td>
        <td>
          Balance
        </td>
        <td>
          Status
        </td>
        <td>
          Edit
        </td>
        <td>
          Delite
        </td>
      </tr>
      <tr
        className="client-row"
      >
        <td>
          Иван
        </td>
        <td>
          Иванов
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="active"
        >
          active
        </td>
        <td>
          <button
            dataTestId="edit1"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
        <td>
          <button
            dataTestId="delite1"
            onClick={[Function]}
          >
            Delite
          </button>
        </td>
      </tr>
      <tr
        className="client-row"
      >
        <td>
          Сидор
        </td>
        <td>
          Сидоров
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="active"
        >
          active
        </td>
        <td>
          <button
            dataTestId="edit2"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
        <td>
          <button
            dataTestId="delite2"
            onClick={[Function]}
          >
            Delite
          </button>
        </td>
      </tr>
      <tr
        className="client-row"
      >
        <td>
          Петр
        </td>
        <td>
          Петров
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="active"
        >
          active
        </td>
        <td>
          <button
            dataTestId="edit3"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
        <td>
          <button
            dataTestId="delite3"
            onClick={[Function]}
          >
            Delite
          </button>
        </td>
      </tr>
      <tr
        className="client-row"
      >
        <td>
          Григорий
        </td>
        <td>
          Григорьев
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="blocked"
        >
          blocked
        </td>
        <td>
          <button
            dataTestId="edit4"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
        <td>
          <button
            dataTestId="delite4"
            onClick={[Function]}
          >
            Delite
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <button
    dataTestId="add"
    onClick={[Function]}
  >
    Add client
  </button>
</div>
`;

exports[`edit buttons work 2`] = `
<div
  className="mobile-company"
>
  <div>
    <button
      onClick={[Function]}
    >
      Velcom
    </button>
    <button
      onClick={[Function]}
    >
      MTS
    </button>
    <br />
    Company: 
    <span>
      Velcom
    </span>
  </div>
  <hr />
  <button
    dataTestId="filterAll"
    onClick={[Function]}
  >
    All
  </button>
  <button
    dataTestId="filterActiv"
    onClick={[Function]}
  >
    Active
  </button>
  <button
    dataTestId="filterBlock"
    onClick={[Function]}
  >
    Blocked
  </button>
  <hr />
  <div
    className="client-card"
  >
    <div
      className="client-card__title"
    >
      Edit existed client
    </div>
    <div
      className="client-card__body"
    >
      <div
        className="client-card__field"
      >
        <span
          className="client-card__form-article"
        >
          ID:
        </span>
        1
      </div>
      <div
        className="client-card__field"
      >
        <span
          className="client-card__form-article"
        >
          First Name:
        </span>
        <input
          className="client-card__input"
          defaultValue="Иван"
          onBlur={[Function]}
          type="text"
        />
        <span
          className="client-card__error"
        />
      </div>
      <div
        className="client-card__field"
      >
        <span
          className="client-card__form-article"
        >
          Last Name:
        </span>
        <input
          className="client-card__input"
          defaultValue="Иванов"
          onBlur={[Function]}
          type="text"
        />
        <span
          className="client-card__error"
        />
      </div>
      <div
        className="client-card__field"
      >
        <span
          className="client-card__form-article"
        >
          Patronymic:
        </span>
        <input
          className="client-card__input"
          defaultValue="Иванович"
          onBlur={undefined}
          type="text"
        />
        <span
          className="client-card__error"
        />
      </div>
      <div
        className="client-card__field"
      >
        <span
          className="client-card__form-article"
        >
          Balance:
        </span>
        <input
          className="client-card__input"
          defaultValue={200}
          onBlur={[Function]}
          type="text"
        />
        <span
          className="client-card__error"
        />
      </div>
      <div
        className="client-card__field"
      >
        <input
          disabled={false}
          onClick={[Function]}
          type="button"
          value="Save"
        />
        <input
          onClick={[Function]}
          type="button"
          value="Cancel"
        />
      </div>
    </div>
  </div>
  <table
    className="clients-table"
  >
    <tbody>
      <tr
        className="table__head"
      >
        <td>
          First Name
        </td>
        <td>
          Last Name
        </td>
        <td>
          Patronymic
        </td>
        <td>
          Balance
        </td>
        <td>
          Status
        </td>
        <td>
          Edit
        </td>
        <td>
          Delite
        </td>
      </tr>
      <tr
        className="client-row"
      >
        <td>
          Иван
        </td>
        <td>
          Иванов
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="active"
        >
          active
        </td>
        <td>
          <button
            dataTestId="edit1"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
        <td>
          <button
            dataTestId="delite1"
            onClick={[Function]}
          >
            Delite
          </button>
        </td>
      </tr>
      <tr
        className="client-row"
      >
        <td>
          Сидор
        </td>
        <td>
          Сидоров
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="active"
        >
          active
        </td>
        <td>
          <button
            dataTestId="edit2"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
        <td>
          <button
            dataTestId="delite2"
            onClick={[Function]}
          >
            Delite
          </button>
        </td>
      </tr>
      <tr
        className="client-row"
      >
        <td>
          Петр
        </td>
        <td>
          Петров
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="active"
        >
          active
        </td>
        <td>
          <button
            dataTestId="edit3"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
        <td>
          <button
            dataTestId="delite3"
            onClick={[Function]}
          >
            Delite
          </button>
        </td>
      </tr>
      <tr
        className="client-row"
      >
        <td>
          Григорий
        </td>
        <td>
          Григорьев
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="blocked"
        >
          blocked
        </td>
        <td>
          <button
            dataTestId="edit4"
            onClick={[Function]}
          >
            Edit
          </button>
        </td>
        <td>
          <button
            dataTestId="delite4"
            onClick={[Function]}
          >
            Delite
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <button
    dataTestId="add"
    onClick={[Function]}
  >
    Add client
  </button>
</div>
`;
